src/core/SampleBuffer.cpp

```diff
@@ -23,1606 +23,109 @@
-
-
-#include <sndfile.h>
-
-#define OV_EXCLUDE_STATIC_CALLBACKS
-#ifdef LMMS_HAVE_OGGVORBIS
-#include <vorbis/vorbisfile.h>
-#endif
-
-#ifdef LMMS_HAVE_FLAC_STREAM_ENCODER_H
-#include <FLAC/stream_encoder.h>
-#endif
-
-#ifdef LMMS_HAVE_FLAC_STREAM_DECODER_H
-#include <FLAC/stream_decoder.h>
-#endif
-
-
-#include "AudioEngine.h"
-#include "base64.h"
-#include "ConfigManager.h"
-#include "DrumSynth.h"
-#include "endian_handling.h"
-#include "Engine.h"
-#include "GuiApplication.h"
-#include "Note.h"
 #include "PathUtil.h"
+#include "SampleDecoder.h"
 
-#include "FileDialog.h"
+namespace lmms {
 
-namespace lmms
+SampleBuffer::SampleBuffer(const sampleFrame* data, int numFrames, int sampleRate)
+	: m_data(data, data + numFrames)
+	, m_sampleRate(sampleRate)
 {
-
-SampleBuffer::SampleBuffer() :
-	m_userAntiAliasWaveTable(nullptr),
-	m_audioFile(""),
-	m_origData(nullptr),
-	m_origFrames(0),
-	m_data(nullptr),
-	m_frames(0),
-	m_startFrame(0),
-	m_endFrame(0),
-	m_loopStartFrame(0),
-	m_loopEndFrame(0),
-	m_amplification(1.0f),
-	m_reversed(false),
-	m_frequency(DefaultBaseFreq),
-	m_sampleRate(audioEngineSampleRate())
-{
-
-	connect(Engine::audioEngine(), SIGNAL(sampleRateChanged()), this, SLOT(sampleRateChanged()));
-	update();
 }
 
-
-
-SampleBuffer::SampleBuffer(const QString & audioFile, bool isBase64Data)
-	: SampleBuffer()
+SampleBuffer::SampleBuffer(const QString& audioFile)
 {
-	if (isBase64Data)
-	{
-		loadFromBase64(audioFile);
-	}
-	else
+	if (audioFile.isEmpty()) { throw std::runtime_error{"Failure loading audio file: Audio file path is empty."}; }
+	auto resolvedFileName = PathUtil::toAbsolute(PathUtil::toShortestRelative(audioFile));
+
+	if (auto decodedResult = SampleDecoder::decode(resolvedFileName))
 	{
+		auto& [data, sampleRate] = *decodedResult;
+		m_data = std::move(data);
+		m_sampleRate = sampleRate;
 		m_audioFile = audioFile;
-		update();
+		return;
 	}
-}
-
-
-
 
-SampleBuffer::SampleBuffer(const sampleFrame * data, const f_cnt_t frames)
-	: SampleBuffer()
-{
-	if (frames > 0)
-	{
-		m_origData = MM_ALLOC<sampleFrame>( frames);
-		memcpy(m_origData, data, frames * BYTES_PER_FRAME);
-		m_origFrames = frames;
-		update();
-	}
+	throw std::runtime_error{
+		"Failed to decode audio file: Either the audio codec is unsupported, or the file is corrupted."};
 }
 
-
-
-
-SampleBuffer::SampleBuffer(const f_cnt_t frames)
-	: SampleBuffer()
+SampleBuffer::SampleBuffer(const QByteArray& base64Data, int sampleRate)
+	: m_data(reinterpret_cast<const sampleFrame*>(base64Data.data()),
+		reinterpret_cast<const sampleFrame*>(base64Data.data()) + base64Data.size() / sizeof(sampleFrame))
+	, m_sampleRate(sampleRate)
 {
-	if (frames > 0)
-	{
-		m_origData = MM_ALLOC<sampleFrame>( frames);
-		memset(m_origData, 0, frames * BYTES_PER_FRAME);
-		m_origFrames = frames;
-		update();
-	}
 }
 
-
-
-
-SampleBuffer::SampleBuffer(const SampleBuffer& orig)
+SampleBuffer::SampleBuffer(const std::vector<sampleFrame>& data, int sampleRate)
+	: m_data(data)
```

Comment:
```suggestion
SampleBuffer::SampleBuffer(std::vector<sampleFrame> data, int sampleRate)
	: m_data(std::move(data))
```
This constructor is making an extra copy of `data`.

I think it's often beneficial for constructors to pass in large types by value instead of const reference when those types define a move constructor, then moving that value into the member variable inside the member initializer list. That way, if the call site uses `std::move()` or a temporary, the move constructor will be picked and you can avoid any copies, but if the call site doesn't do that and just passes the argument normally, the copy constructor will be picked and you'll still only have one copy just like before.